server/auth.go

```diff
@@ -53,6 +70,14 @@ func (p *Permissions) clone() *Permissions {
 	return clone
 }
 
+func (s *Server) SetCustomClientAuth(auth Auth) {
+	s.customClientAuth = auth
```

Comment:
I believe that we should protect this and make sure one can set it only before `server.Start()` has been invoked and return and error otherwise. 
Without locking/protection, any one could call this function (same for `SetCustomRouterAuth`) at any point when server is running and that would produce a race if code is trying to authorize a client.

One approach would be to protect setting this with `s.mu.Lock()/Unlock()` and check for `s.running`. If true, it means `Start()` has been invoked and therefore this call should fail (should return `error`).

A test case using those 2 new APIs (with dummy custom auth) would be welcome, including situations trying to set custom auth after server has been started.